### ISSUE TITLE
DB023 - Crazy amount of dbListenerTokens - makes the DB run slow

### ISSUE THREAD

**Issue by itaishalom** [couchbase/couchbase-lite-android#1693]
Hi,
I recently encountered an issue, db operations become very slow.
We looked into the Database object and we found out the following - 
We use liveQueryListner, on regular case there are 3 of those (we saw on activeLiveQuery field) and there are 6 dbListnereTokens, two for each live query (with the same toString..)

But sometimes on a different scenario there are 22 activeLiveQueries but 6867 dbListenersTokens,
So the first question is why there are so many dbListenersTokens? why the DB runs so slow?

Edit - Further investigating reveals that there are 2 additional dbListenerTokens after each time that change() is called (on any livequery). We use @hideki 's code to update the live query params - maybe this causes additional tokens?
Thanks

**Comment by hideki** [comment]
HI @itaishalom,
Can you share the sample app that can reproduce this issue?

**Comment by hideki** [comment]
PR: https://github.com/couchbase/couchbase-lite-android/pull/1694

**Comment by hideki** [comment]
How to reproduce:

1. Clone Android project from 
- https://github.com/itaishalom/CouchBase/tree/newTest
- https://github.com/itaishalom/couchbaseClientTest
https://github.com/couchbase/couchbase-lite-android/issues/1628#issuecomment-371586622

2. Apply patch
```
public class MyQueryListner {
    private ListenerToken mToken;
    Query mQuery;
    long lastTime = 0;
    myQueryListnerImpl myQueryListner;
    //public static int counter = 0;
    AtomicInteger counter = new AtomicInteger(0);

    public MyQueryListner() {
        mQuery = QueryBuilder.select(SelectResult.all())
                .from(DataSource.database(DBHandler.getInstance().mDb))
                .where(Expression.property(TIME).greaterThan(Expression.parameter(TIME)))
                .orderBy(Ordering.property(TIME));
        Parameters params = new Parameters(mQuery.getParameters());
        params.setValue(TIME, lastTime);
        mQuery.setParameters(params);
        myQueryListner = new myQueryListnerImpl();
        mToken = mQuery.addChangeListener(myQueryListner);
    }

    public class myQueryListnerImpl implements QueryChangeListener {
        @Override
        public void changed(QueryChange change) {
            ResultSet rows = change.getResults();
            if (rows == null)
                return;
            Result row;
            boolean rowsExists = false;

            while ((row = rows.next()) != null) {
                //counter++;
                counter.incrementAndGet();
                rowsExists = true;
                //Log.w("MyQueryListner", row.toMap().toString());
                lastTime = row.getDictionary("testDB").getLong(TIME);
            }

            if (rowsExists) {
                //lastTime = System.currentTimeMillis();
                Parameters params = new Parameters(mQuery.getParameters());
                params.setValue(TIME, lastTime);
                mQuery.setParameters(params);
                Log.w("MyQueryListner", "insert: notifications: " + counter.get());
            }
        }
    }
}
```
https://github.com/couchbase/couchbase-lite-android/issues/1628#issuecomment-373886433

3. Press `Test` button

4. Count number of `DatabaseChangeListenerToken` instance with using Heap Dump

**Comment by hideki** [comment]
@itaishalom 
Could you please verify the fix with `2.0.0-457`?
http://mobile.maven.couchbase.com/feeds/dev//com/couchbase/lite/couchbase-lite-android/2.0.0-457

**Comment by itaishalom** [comment]
Wonderful! Great fix, thank you very much
